### PR TITLE
Support mutually recursive types

### LINKED ISSUE #45: Mutually recursive types
It seems that single recursive types are supported since #31, but mutually recursive types using `and` are not.

Repro:

```reason
[@decco] type inttree =
  | Empty
  | Node(node)
[@decco] and node = {
  value: int,
  left: inttree,
  right: inttree,
};
```

It triggers an error `The value node_encode can't be found`.

I believe that encoders/decoders should also be declared with `and` so they can call each other.

### PR DESCRIPTION
**Pull request by jchavarri**
Fixes #45.

The change consists on keeping 1:1 relationship between structure items. So for each structure item with the `@decco` attribute, another one will be created. If the original structure item had many types in it, the newly created will have multiple value bindings for the encoders / decoders as well.

~This means that there is no need to annotate each type in a mutually recursive type declaration. Just doing~:

```reason
[@decco] type inttree =
    | Empty
    | Node(node)
and node = {
    value: int,
    left: inttree,
    right: inttree,
};
```

~Will generate encoder / decoder for `node` as well.~

Edit: fake news 😅 Each type annotation still needs to be decorated with `@decco`:

```reason
[@decco] type inttree =
    | Empty
    | Node(node)
[@decco] and node = {
    value: int,
    left: inttree,
    right: inttree,
};
```

The recursive flag logic got simplified as well, the encoders/decoders structure item gets the same flag as the original type definition. Because recursive is the default, and most of the times the type definition is not really recursive, the warning 39 in generated code has been muted as well.

### CODE REVIEW

File: ppx_src/bin/Structure.re [hunk 1]
```diff
@@ -23,7 +23,15 @@ let generateCodecDecls = (typeName, paramNames, (encoder, decoder), isRecursive)
 
     let vbs = switch encoder {
     | None => vbs
-    | Some(encoder) => vbs @ [Vb.mk(encoderPat, addParams(encoderParamNames, encoder))]
+    | Some(encoder) =>
+      vbs
+      @ [
+        Vb.mk(
+          ~attrs=[attrWarning([%expr "-39"])],
```
Comment (ryb73) [inline ppx_src/bin/Structure.re]: nice 👍

**Comment by DCKT** [comment]
I was about to create an issue asking for this feature ! Is this mergeable ? It could be so great !

**Comment by bnoguchi** [comment]
@jchavarri great work!

@ryb73 thanks for creating and supporting decco! Looking forward to this merge and a new release.

**Comment by jchavarri** [comment]
@ryb73 I got distracted 😅 I added a couple of tests, could you take a look, please?

**Comment by jchavarri** [comment]
Not sure why GH didn't pick up the tests result but it passed: https://travis-ci.org/github/reasonml-labs/decco/builds/668041360.

**Comment by bnoguchi** [comment]
@ryb73 any plans to merge this in and publish a new release to npm soon?

**Comment by ryb73** [comment]
Sorry, been really busy and haven't been working on Reason lately. I'm
hoping to have time to release this by the weekend.

**Comment by ryb73** [comment]
This is awesome, thanks @jchavarri ! Sorry for the delay

**Comment by bnoguchi** [comment]
Thanks @ryb73 !

**Pull request merged**